File: tests/e2e/cases/06-int-ops/popcount_u8/popcount_u8_disproved.rs
// DEMO: BUGGY 8-bit popcount (Rust mirror) — stops at i < 7, misses bit 7.
// SAW disproves with x = 0x80.

fn popcount_u8(x: u8) -> u32 {
    let mut n: u32 = 0;
    let mut i: u32 = 0;
    while i < 7 {   // BUG: was i < 8.
        n += ((x >> i) & 1) as u32;
        i += 1;
    }
    n
}

File: tests/e2e/cases/06-int-ops/popcount_u8/popcount_u8_disproved.cpp
/*
DEMO: BUGGY 8-bit popcount — off-by-one on the loop bound.

Stops at i < 7 instead of i < 8, missing the top bit. SAW disproves
with x = 0x80 (only the missed bit set): impl returns 0, spec says 1.

Expected verdict: DISPROVED.
*/

#include <cstdint>

uint32_t popcount_u8(uint8_t x) {
    uint32_t n = 0;
    for (uint32_t i = 0; i < 7; i++) {   // BUG: was i < 8.
        n += static_cast<uint32_t>((x >> i) & 1u);
    }
    return n;
}
